core/src/solution_submission.rs

```diff
@@ -174,77 +174,105 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
-    let mut gas_price_factor = 1;
-    let mut result;
-    // the following block emulates a do-while loop
-    while {
-        gas_price_estimate = match gas_price_estimating.estimate_gas_price().await {
-            Ok(gas_estimate) => gas_estimate.fast,
-            Err(ref err) => {
-                log::warn!(
-                    "failed to get gas price from gnosis safe gas station: {}",
-                    err
-                );
-                gas_price_estimate
-            }
-        };
-        // Never exceed the gas cap.
-        let gas_price = std::cmp::min(gas_price_estimate * gas_price_factor, gas_cap);
+    const MIN_GAS_PRICE_INCREASE_FACTOR: f64 = 1.125 * (1.0 + f64::EPSILON);
+
+    let effective_gas_price_cap = U256::from(
+        (gas_price_cap.as_u128() as f64 / MIN_GAS_PRICE_INCREASE_FACTOR).floor() as u128,
+    );
+    assert!(effective_gas_price_cap <= gas_price_cap);
+
+    let mut gas_price_estimator =
+        InfallibleGasPriceEstimator::new(gas_price_estimating, DEFAULT_GAS_PRICE.into());
+
+    let submit_solution = |gas_price, block_timeout| {
+        contract.submit_solution(
+            batch_index,
+            solution.clone(),
+            claimed_objective_value,
+            gas_price,
+            block_timeout,
+        )
+    };
 
-        // Submit solution at estimated gas price (not exceeding the cap)
-        result = contract
-            .submit_solution(
-                batch_index,
-                solution.clone(),
-                claimed_objective_value,
-                gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
-            )
-            .await;
-
-        let gas_price_can_still_increase = gas_price < effective_gas_cap;
-        // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
-    } {
-        // Increase gas
-        gas_price_factor *= INCREASE_FACTOR;
+    for gas_price_increase_count in 0u32.. {
+        let estimated_price = gas_price_estimator.estimate().await;
+        let gas_price = gas_price(estimated_price, gas_price_increase_count, gas_price_cap);
+        assert!(gas_price <= gas_price_cap);
         info!(
-            "retrying solution submission with increased gas price factor of {}",
-            gas_price_factor,
+            "solution submission try {} with gas price {}",
+            gas_price_increase_count, gas_price
         );
+        let is_last_iteration = gas_price >= effective_gas_price_cap;
+        let block_timeout = if is_last_iteration {
+            None
+        } else {
+            Some(BLOCK_TIMEOUT)
+        };
+        let result = submit_solution(gas_price, block_timeout).await;
```

Comment:
ubernit: I don't think the closure has so much value anymore seeing as this is the only place the `submit_solution` is called.